bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
+        *self = self.wrapping_sub(1);
+    }
+
+    fn set_value(&mut self, value: u8) {
+        *self = value;
+    }
+
+    fn get_value(&self) -> u8 {
+        *self
+    }
+}
+
+/// Error types that the [VirtualMachine] can emit
+#[derive(Debug, PartialEq, Eq)]
+pub enum VMError {
+    /// The head ran off the start of the tape
+    HeadUnderrun(LocalisedInstruction),
+    /// The head ran off the end of the (non-auto-extending) tape
+    HeadOverrun(LocalisedInstruction),
+    /// Reading a byte from stdio went bloop
+    ReadError(LocalisedInstruction, String),
+    /// Writing a byte from stdio went bloop
+    WriteError(LocalisedInstruction, String),
+}
+
+impl std::error::Error for VMError {}
+
+impl Display for VMError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::HeadOverrun(program_instruction) => {
+                write!(
+                    f,
+                    "Head Overrun Error occured at line {} column {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num()
+                )
+            }
+            Self::HeadUnderrun(program_instruction) => {
+                write!(
+                    f,
+                    "Head Underrun Error occured at line {} column {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num()
+                )
+            }
+            Self::ReadError(program_instruction, error) => {
+                write!(
+                    f,
+                    "Read Error occured at line {} column {}: {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num(),
+                    error
+                )
+            }
+            Self::WriteError(program_instruction, error) => {
+                write!(
+                    f,
+                    "Write Error occured at line {} column {}: {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num(),
+                    error
+                )
+            }
+        }
+    }
+}
+
+impl From<(LocalisedInstruction, std::io::Error)> for VMError {
+    fn from(value: (LocalisedInstruction, std::io::Error)) -> Self {
+        let bad_instruction = value.0;
+        let error_msg = value.1.to_string();
+        if bad_instruction.instruction() == Instruction::Input {
+            VMError::ReadError(bad_instruction, error_msg)
+        } else {
+            VMError::WriteError(bad_instruction, error_msg)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::{ErrorKind, Write};
+    use tempfile::NamedTempFile;
+
+    fn test_program() -> BfProgram {
+        let mut test_file = NamedTempFile::new().unwrap();
+        test_file.write_all(b",.[test]+++.").unwrap();
+
+        BfProgram::from_file(test_file.path()).unwrap()
+    }
```

Comment:
Use of `NamedTempFile` like this is certainly interesting.  Are you doing this deliberately to test `from_file()` ?  If not, then you're doing an awful lot of IO for something which could simply be:

```suggestion
    fn test_program() -> BfProgram {
        BfProgram::new("testfile.bf", ",.[test]+++.")
    }
```